src/streamlink/utils/formatter.py

```diff
@@ -0,0 +1,64 @@
+import re
+
+from streamlink_cli.compat import is_win32
+from streamlink_cli.constants import DEFAULT_STREAM_METADATA, FS_SAFE_REPLACEMENT_CHAR, INVALID_FILENAME_CHARS
+
+
+class Formatter(dict):
+    def __init__(self, plugin, args):
+        self.plugin = plugin
+        self.args = args
+        self.cache = self.Cache(self.plugin)
+
+    def __missing__(self, k):
+        if k in ("title", "author", "category", "game", "url"):
```

Comment:
seems hard coded, `Formatter` is not really reusable for other stuff

Reply:
That is true, but it was never intended to be a general reusable formatter.  Its purpose is only to supply formatted title and filesystem-safe filename strings in a lazy manner and without multiple calls to the `plugin.get_*()` metadata methods.  Perhaps I should rename it to better reflect that.